web/packages/teleport/src/lib/polyfillRandomUuid.js

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+if (typeof crypto?.randomUUID != 'function') {
+  window.crypto.randomUUID = randomUUID;
+}
+
+// Code based on Node.js' `lib/internal/crypto/random.js`, subject
+// to Node.js license found at:
+// https://raw.githubusercontent.com/nodejs/node/master/LICENSE
+
+//
+// internal/errors
+//
+class ERR_INVALID_ARG_TYPE extends TypeError {
+  constructor(name, type, value) {
+    super(`${name} variable is not of type ${type} (value: '${value}')`);
+  }
+
+  code = 'ERR_INVALID_ARG_TYPE';
+}
+
+//
+// internal/validators
+//
+
+function validateBoolean(value, name) {
+  if (typeof value !== 'boolean')
+    throw new ERR_INVALID_ARG_TYPE(name, 'boolean', value);
+}
+
+function validateObject(value, name) {
+  if (value === null || Array.isArray(value) || typeof value !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(name, 'Object', value);
+  }
+}
+
+//
+// crypto
+//
+
+const randomFillSync =
+  typeof window === 'undefined'
+    ? require('crypto').randomFillSync
+    : window.crypto.getRandomValues.bind(window.crypto);
```

Comment:
[The docs for `crypto.getRandomValues`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) say this:

> The `Crypto.getRandomValues()` method lets you get cryptographically strong random values. The array given as the parameter is filled with random numbers (random in its cryptographic meaning).
>
> To guarantee enough performance, implementations are not using a truly random number generator, but they are using a pseudo-random number generator seeded with a value with enough entropy. The pseudo-random number generator algorithm (PRNG) may vary across [user agents](https://developer.mozilla.org/en-US/docs/Glossary/User_agent), but is suitable for cryptographic purposes.
>
> `getRandomValues()` is the only member of the `Crypto` interface which can be used from an insecure context.

The part about being suitable for cryptographic purposes is good to hear. I was afraid that we're potentially replacing a cryptographically secure `crypto.randomUUID` with an insecure version. While I'm not sure if this property is needed for what we use `crypto.randomUUID` for, it would potentially mean that 3rd party libs that expect this property to hold would suddenly operate on an insecure version.

However, do you know why `crypto.randomUUID` is available only from secure contexts? If so, do we risk anything by polyfilling `crypto.randomUUID` even on HTTP pages?

Reply:
Based on the discussion in https://github.com/WICG/uuid/issues/23, it doesn't seem like the decision to make `crypto.randomUUID` only available from secure contexts was made due to any specific implementation concerns, and it seems to have just been a general design decision to expose new API's in secure contexts only. I don't think we risk anything by polyfilling it in insecure contexts, especially given that none of our implementations of it are security-sensitive.